src/scanspec/core.py

```diff
@@ -104,134 +102,106 @@ def calculate(self) -> int:
 
-T = TypeVar("T", type, Callable)
+    super_cls.__init_subclass__ = classmethod(add_subclass_to_union)  # type: ignore
+    super_cls.__get_pydantic_core_schema__ = classmethod(get_schema_of_union)  # type: ignore
+    return super_cls
 
 
-def deserialize_as(cls, obj):
-    return _tagged_unions[cls].type_adapter.validate_python(obj)
+T = TypeVar("T", type, Callable)
 
 
 def uses_tagged_union(cls_or_func: T) -> T:
     """
-    Decorator that processes the type hints of a class or function to detect and
-    register any tagged unions. If a tagged union is detected in the type hints,
-    it registers the class or function as a referrer to that tagged union.
-    Args:
-        cls_or_func (T): The class or function to be processed for tagged unions.
-    Returns:
-        T: The original class or function, unmodified.
+    T = TypeVar("T", type, Callable)
+        Decorator that processes the type hints of a class or function to detect and
+        register any tagged unions. If a tagged union is detected in the type hints,
+        it registers the class or function as a referrer to that tagged union.
+        Args:
+            cls_or_func (T): The class or function to be processed for tagged unions.
+        Returns:
+            T: The original class or function, unmodified.
     """
-    for k, v in get_type_hints(cls_or_func).items():
+    for v in get_type_hints(cls_or_func).values():
         tagged_union = _tagged_unions.get(get_origin(v) or v, None)
         if tagged_union:
-            tagged_union.add_referrer(cls_or_func, k)
+            tagged_union.add_reference(cls_or_func)
     return cls_or_func
 
 
+_tagged_unions: dict[type, _TaggedUnion] = {}
+
+
 class _TaggedUnion:
     def __init__(self, base_class: type, discriminator: str):
         self._base_class = base_class
-        # The members of the tagged union, i.e. subclasses of the baseclasses
-        self._members: list[type] = []
         # Classes and their field names that refer to this tagged union
-        self._referrers: dict[type | Callable, set[str]] = {}
-        self.type_adapter: TypeAdapter = TypeAdapter(None)
         self._discriminator = discriminator
-
-    def _make_union(self):
-        if len(self._members) > 0:
-            return Union[tuple(self._members)]  # type: ignore  # noqa
-
-    def _set_discriminator(self, cls: type | Callable, field_name: str, field: Any):
-        # Set the field to use the `type` discriminator on deserialize
-        # https://docs.pydantic.dev/2.8/concepts/unions/#discriminated-unions-with-str-discriminators
-        if isclass(cls):
-            assert isinstance(
-                field, FieldInfo
-            ), f"Expected {cls.__name__}.{field_name} to be a Pydantic field, not {field!r}"  # noqa: E501
-            field.discriminator = self._discriminator
+        # The members of the tagged union, i.e. subclasses of the baseclass
+        self._members: list[type] = []
```

Comment:
Nit: `members` is an overloaded term, maybe call this `self._subclasses` to be explicit?